### PR TITLE
Logger format improvements

### PR DESCRIPTION
**Pull request by neclimdul**
Key changes:
- Some generalization in file name generation to allow future customization.
- Use the DateTime class and add ".u" to format to prevent loss of request logs from the same second.

### CODE REVIEW

**Comment by neclimdul** [comment]
Found this trying to figure out why sometimes there where missing requests in a series of paged search requests I was debugging. Though there was a bug in my paging logic but just turned out sometimes netsuite was being too quick and triggering multiple requests in the same second leaving missing log files.

While I was poking around, I've wanted to change the format around a bit rather the being stuck with the out of the box file names. Sometimes I have to pass some logs off to debug things and I'm always having to explain who ryan is :-D

I added the code to customize the log file, however I wasn't confident in how to wire it up. Need to look back at #63 and make sure it doesn't break the BC left in place there. That can easily be a follow up though.

**Comment by ryanwinchester** [comment]
Haha, my name was in the logfiles, too? What was I thinking? 😂

**Comment by neclimdul** [comment]
@ryanwinchester lol, well that's an easy fix I can include. makes sense with it being in an org now. ;)

**Comment by jrebs** [comment]
Hi James, I was a bit unsure how to submit a PR to your personal branch for review since I already have the main netsuite-php repo forked and so I just pushed the commit right to your branch not realizing that's what it was going to do, exactly. :/

For your review, the following additions to your re-working of the logging functionality:
* Allow newly added `log_format` and `log_dateformat` to be defined as configuration values
* Pass said values into the new Logger() object when present or else use your defaults
* Add `Logger::setPath()` so that `NetSuiteClient::setLogPath()` can modify the existing Logger object instead of instantiating a new one
* Expand the README and default .env file to document all the new logger options

**Comment by neclimdul** [comment]
I think that all looks right but I don't understand why you changed the constructor logic

**Comment by jrebs** [comment]
Ah, good eye. I should have explained that. The reasoning for that was that I know some people (myself included) incorporate this library with laravel applications and in those applications, it's not uncommon to see configuration keys defaulted to empty strings or nulls when they have not been provided by the user at runtime. An example from a `config/netsuite.php` file might look like this:

```php
return [
  // ...
  'log_format' => env('NETSUITE_LOG_FORMAT', ''),
  // ...
]
```

The net result being that if the user doesn't declare this key in their application setup, then the configuration array passed into the netsuite-php library will include all of the possible keys with empty strings or nulls for ones that the user did not provide.

The constructor logic as it was written would only set the default constant value when the configuration value was not passed in at all, but if a null or empty string was passed in, it would treat it as the desired value. For systems like the above, this would make it a bit fussy to allow for the defaults to be used, as you'd need to prevent anything from being passed in at all, and I was going on the assumption that we wouldn't want to force the user to manually instantiate the logger object to override these values.

A basic demonstration:

```php
class Test {
  const DEFAULT_FOO = 'foo';

  public __construct($foo = self::FOO)
  {
    var_dump($foo);
  }
}

new Test();
new Test(null);
new Test('');
new Test('bar');

// string(3) "foo"
// NULL
// string(0) ""
// string(3) "bar"
```

My thinking was that the desired functionality here would be for the default values to be used any time an explicit alternate wasn't provided, and to make this possible through configuration alone.

I'm certainly not of the mind that all of the above is the Right Way for this to be done and that's why I wanted to chew on it with you. TBH I had intended to just write this as a PR to propose but I botched that. :/

**Comment by neclimdul** [comment]
Have you ever been like "man this needs a simple review" but then you put it off for like 4 months because you'd like to provide some code and keep getting distracted?

I too use this library with Laravel. Basically exclusively :rofl:  Maybe this would help? #88

I haven't tested that exact provider too much but I have an internal library with some helpers and its basically the exact version I have in it and its been working great for a very long time. 

I believe with this provider the logger would never pass anything to the 2nd and 3rd logger options (since they're set in the constructor) and defaults would be setup the same before your changes.

Its hard to remember anymore but I believe my intention was to support modifying those through the provider, laravel config, and the client config array. However, it doesn't seem like I setup the NetSuiteClient constructor to support that so I'm a little confused. Maybe I was trying to keep my changes to a minimum? Probably just a lazy mistake if I'm honest.

**Comment by neclimdul** [comment]
rebased without the constructor changes.

**Comment by jrebs** [comment]
Hi James,

I gave this a test and it refreshed my memory as to what I thought was wrong about it previously. If you do not define the `log_format` and/or `log_dateformat` config keys, those values end up being unset values in the `NetSuiteClient::construct()` method. Then at line 66 the `Logger` object is created and since those values are unset, it pases `null` values into `Logger::construct()`. The way you have defined the constructor is such that the null values passed in are being accepted and used as nulls. The default values you have set in the constructor parameters are only used if the value is not passed in at all (which would never be the case). This is why in the last iteration, I had changed the constructor to look at the values in the constructor and then used the class constant defaults when it found nulls. I will let you decide how to address this issue rather than cross wires, but as it stands now, if a user chooses not to define the those keys, they end up with log files called `-request.xml` and `-response.xml` which overwrite themselves with every call.

Less important, but on a separate note I would like to try to move towards [PSR-12 conformity](https://www.php-fig.org/psr/psr-12/) with the project as a coding standard. Strict adherence is not required and I expect the move to be gradual, but I thought I'd note it know, as there are some things that stood out to me in this review which are candidate to change (such as NULL => null, etc).

**Comment by neclimdul** [comment]
Ah, I think I see what you're saying now. It should have been something like this?

```php
        $this->logger = new Logger(
            empty($this->config['log_path']) ? $this->config['log_path'] : NULL,
            empty($this->config['log_format']) ? $this->config['log_format'] : Logger::DEFAULT_LOG_FORMAT,
            empty($this->config['log_dateformat']) ? $this->config['log_dateformat'] : Logger::DEFAULT_DATE_FORMAT
        );
```

**Comment by jrebs** [comment]
> Ah, I think I see what you're saying now. It should have been something like this?
> 
> ```
>         $this->logger = new Logger(
>             empty($this->config['log_path']) ? $this->config['log_path'] : NULL,
>             empty($this->config['log_format']) ? $this->config['log_format'] : Logger::DEFAULT_LOG_FORMAT,
>             empty($this->config['log_dateformat']) ? $this->config['log_dateformat'] : Logger::DEFAULT_DATE_FORMAT
>         );
> ```

The suggestion does functionally fix the problem, but IMO this is a bit busy looking for an instantiation call? It seems to me like something that another developer would look at in the future and have to spend longer evaluating visually than is necessary. I'm also wondering if it's better to leave the knowledge of the Logger default constants within the Logger class itself, reducing any outside information that other classes need to have about it?

I'd propose something akin to the following:

```php
// In NetSuiteClient::__construct()
$this->logger = new Logger(
    $this->config['log_path'] ?? null,
    $this->config['log_format'] ?? null,
    $this->config['log_dateformat'] ?? null
);
```

```php
class Logger
{
    public function __construct($logPath, $logFormat, $logDateFormat)
    {
        $this->logPath = $logPath ?: null;
        $this->logFormat = $logFormat ?: self::DEFAULT_LOG_FORMAT;
        $this->logDateFormat = $logDateFormat ?: self::DEFAULT_DATE_FORMAT,
    }
}
```

My reasoning:
* The call to instantiate the Logger has a much lower visual load while reading the code
* The null coalesce operators are just to avoid `E_NOTICE` if the user's config is missing these keys entirely, which seems to be likely to occur since these are new keys and many people will have been using the same config for a long time
* The Logger assigns default values if necessary without other classes having to know how

**Comment by neclimdul** [comment]
1 & 2, yeah its a bit busy but I went with empty because it would catch false and empty string since getenv() can be a bit weird. For example, if you populated config out of environment and the format wasn't defined the value would be FALSE which ?? would treat as set and pass through. Kinda error prone.

3, I think the default values already accomplish this. Also, I think the optional parameter method is a bit better because IDE's will show the values for optional parameters values to developers but a developer would have to inspect the logger class to find out the default value logic.

I guess looking at them together it was the idea to catch some falsy values in the client constructor and the rest in the logger constructor? That seems more complex. I definitely think all the special quirky cases specific to the config array/environment magic should be in one place, the client constructor, and there shouldn't be any logic in the logger constructor. That more closely aligns with IoC and makes injection of the logger directly easier in the future.

**Comment by jrebs** [comment]
> 1 & 2, yeah its a bit busy but I went with empty because it would catch false and empty string since getenv() can be a bit weird. For example, if you populated config out of environment and the format wasn't defined the value would be FALSE which ?? would treat as set and pass through. Kinda error prone.

It's correct that `false` and ` `  would make it into the constructor, but that's why inside the Logger constructor I used the short ternary assignment `?:`. This would then turn those false or empty strings (or nulls) into the intended default.

> 3, I think the default values already accomplish this. Also, I think the optional parameter method is a bit better because IDE's will show the values for optional parameters values to developers but a developer would have to inspect the logger class to find out the default value logic.
> 
> I guess looking at them together it was the idea to catch some falsy values in the client constructor and the rest in the logger constructor? That seems more complex. I definitely think all the special quirky cases specific to the config array/environment magic should be in one place, the client constructor, and there shouldn't be any logic in the logger constructor. That more closely aligns with IoC and makes injection of the logger directly easier in the future.

Fair points. My primary concern was making sure that someone with missing configuration keys would still have a functional experience and this should solve it. I'll look over the rest of the PR and I should be able to give it some testing tomorrow before merging it in.

File: README.md
Comment (jrebs) [inline README.md:243]: Looks like the logging section from the previous version of the readme snuck back in here.

**Review by jrebs — CHANGES_REQUESTED** [review]

**Comment by neclimdul** [comment]
I thought maybe I'd missed something writing tests for this but now I remember there aren't tests because I'm not really familiar with spec and it looks kinda hard to reverse a spec into an existing implementation. 

Anyways if you're happy with this is probably OK to merge.

**Comment by jrebs** [comment]
> Anyways if you're happy with this is probably OK to merge.

Thanks, James. I don't think there's anyone waiting on this merge, so I'll let it sit for a moment and merge it on then ext day when I'm working with my NetSuite integrations and I'll run it to make sure nothing else crops up.

**Pull request merged**
